@here/harp-omv-datasource/lib/OmvDecodedTileEmitter.ts

```diff
@@ -730,6 +714,117 @@ export class OmvDecodedTileEmitter implements IOmvEmitter {
+            aLine: number[],
+            startIndex: number,
+            endIndex: number
+        ): number => {
+            let minX = Number.MAX_SAFE_INTEGER;
+            let maxX = Number.MIN_SAFE_INTEGER;
+            let minY = Number.MAX_SAFE_INTEGER;
+            let maxY = Number.MIN_SAFE_INTEGER;
+            for (let i = startIndex; i < endIndex; i += 3) {
+                const x = aLine[i];
+                const y = aLine[i + 1];
+                if (x < minX) {
+                    minX = x;
+                }
+                if (x > maxX) {
+                    maxX = x;
+                }
+                if (y < minY) {
+                    minY = y;
+                }
+                if (y > maxY) {
+                    maxY = y;
+                }
+            }
+
+            return (maxX - minX) * (maxX - minX) + (maxY - minY) * (maxY - minY);
+        };
+
+        // Work on a copy of the path.
+        const pathsToCheck = lines.slice();
+
+        while (pathsToCheck.length > 0) {
+            const path = pathsToCheck.pop();
+
+            if (path === undefined || path.length < 6) {
+                continue;
+            }
+
+            let splitIndex = -1;
+
+            for (let i = 0; i < path.length - 3; i += 3) {
+                tempP0.set(path[i], path[i + 1]);
+                tempP1.set(path[i + 3], path[i + 4]);
+                const tangent = tempP1.sub(tempP0).normalize();
+
+                if (i > 0) {
+                    const theta = Math.atan2(
+                        tempPreviousTangent.x * tangent.y - tangent.x * tempPreviousTangent.y,
+                        tangent.dot(tempPreviousTangent)
+                    );
+
+                    if (Math.abs(theta) > maxCornerAngle) {
+                        splitIndex = i;
+                        break;
+                    }
+                }
+                tempPreviousTangent.set(tangent.x, tangent.y);
+            }
+
+            if (splitIndex > 0) {
+                // Estimate if the first part of the path is long enough for the label.
+                const firstPathLengthSqr = computeBoundingBoxSizeSqr(path, 0, splitIndex + 3);
+                // Estimate if the second part of the path is long enough for the label.
+                const secondPathLengthSqr = computeBoundingBoxSizeSqr(
+                    path,
+                    splitIndex,
+                    path.length
+                );
+
+                if (firstPathLengthSqr > minEstimatedLabelLengthSqr) {
+                    // Split off the valid first path points with a clone of the path.
+                    validLines.push(path.slice(0, splitIndex + 3));
+                }
+
+                if (secondPathLengthSqr > minEstimatedLabelLengthSqr) {
+                    // Now process the second part of the path, it may have to be split
+                    // again.
+                    pathsToCheck.push(path.slice(splitIndex));
```

Comment:
Would it not be more efficient to store an index into `paths`, as opposed to copying the entire array at this point.

Reply:
It may, but it would probably make the algorithm more complicated. I would rather leave it like it is.

Reply:
... and these array functions are really fast....

Reply:
OK, I measured a sample view: (many ridiculously jaggy rivers)
https://render-verity.jenkins.release.in.here.com/job/verity-sdk-pre-submit/6680/Try_Me/mapview-demo/?Latitude=50.08683514&Longitude=16.018397&Altitude=321943.81113998&ZoomLevel=10&Camera%20Height=321943.81113998&Near%20Plane=32194.381114&Far%20Plane=1609919.0556999&Theme=berlinBase&Select%20Style=tilezen

Decoding 20 tiles: 1735ms
Thereof line features: 1005ms
Projecting coordinates: 33ms
Splitting jaggy lines: 27ms

So, splitting the jaggy lines takes around 2.7% of processing the lines, pretty much worst case, which should be changed on the data side anyway.

In this view (All of Germany and more, at level 7):
https://render-verity.jenkins.release.in.here.com/job/verity-sdk-pre-submit/6680/Try_Me/mapview-demo/?Latitude=51.28991288&Longitude=10.66596571&Altitude=1699229.62249431&ZoomLevel=7.6&Camera%20Height=1699229.62249431&Near%20Plane=169922.96224943&Far%20Plane=8496348.11247155&Theme=berlinBase&Select%20Style=tilezen

Splitting takes 2.5ms out of the 360ms used for lines.

New York, level 13, MANY short lines:
https://render-verity.jenkins.release.in.here.com/job/verity-sdk-pre-submit/6680/Try_Me/mapview-demo/?Latitude=40.70600511&Longitude=-73.90617049&Altitude=34885.64662481&ZoomLevel=13.20610344&Camera%20Height=34885.64662481&Near%20Plane=3488.56466248&Far%20Plane=174628.23312405&Theme=berlinBase&Select%20Style=tilezen

Splitting takes 0.6ms (!!!) out of the 1073ms used for lines.

I guess we are good, performance is OK.

Thank you very much for considering performance!!! Love it.

Reply:
Thanks for taking a look.